src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsModel.cs

```diff
@@ -574,34 +680,69 @@ private void AddSchema(CsdlSchema schema, bool addAnnotations)
             {
                 RegisterElement(container);
             }
+        }
 
-            if (!string.IsNullOrEmpty(schema.Alias))
+        private void AddOutOfLineAnnotationsFromSchema(
+            CsdlSchema schema,
+            CsdlSemanticsSchema schemaWrapper,
+            Dictionary<string, List<IEdmIncludeAnnotations>> includeAnnotationsIndex)
+        {
+            foreach (CsdlAnnotations schemaOutOfLineAnnotations in schema.OutOfLineAnnotations)
             {
-                this.SetNamespaceAlias(schema.Namespace, schema.Alias);
+                string target = this.ReplaceAlias(schemaOutOfLineAnnotations.Target);
+                
+                List<CsdlSemanticsAnnotations> annotations;
+                if (!this.outOfLineAnnotations.TryGetValue(target, out annotations))
+                {
+                    annotations = new List<CsdlSemanticsAnnotations>();
+                    this.outOfLineAnnotations[target] = annotations;
+                }
+
+                CsdlAnnotations filteredAnnotations = FilterIncludedAnnotations(schemaOutOfLineAnnotations, target, includeAnnotationsIndex);
+                annotations.Add(new CsdlSemanticsAnnotations(schemaWrapper, filteredAnnotations));
             }
+        }
 
-            if (addAnnotations)
+        private CsdlAnnotations FilterIncludedAnnotations(
+            CsdlAnnotations csdlAnnotations,
+            string target,
+            Dictionary<string, List<IEdmIncludeAnnotations>> includeAnnotationsIndex)
+        {
+            if (includeAnnotationsIndex == null)

```

Comment:
Is this ever likely to evaluate to `true` given the initialization of `includeAnnotationsIndex` [here](https://github.com/OData/odata.net/pull/1987/files#diff-f5d2f568fee616e97aad18909cd5716484d0e0b873569410006f6896b0c5c6f4R602)? Maybe check for `null` or empty (or just empty if `null` is unlikely)?

Reply:
It's actually set to null when called from the original `AddSchema(schema, addAnnotations)` method [here](https://github.com/OData/odata.net/pull/1987/files#diff-f5d2f568fee616e97aad18909cd5716484d0e0b873569410006f6896b0c5c6f4R535-R538).

In that method, when the `addAnnotations` was set true, all the annotations in the schema were added without any filtering. In that case I call the `AddSchemaOutOfLineAnnotations` with `includeAnnotationsIndex` set to `null` to indicate that all annotations should be added without any filtering. If, on the other hand, the `includeAnnotationsIndex` is empty, then all annotations will be filtered out and not get added, since the `ShouldIncludeAnnotation` method will return false.

I think I should mention in the doc comments what the intended behaviour is when `includeAnnotationsIndex` is `null`.

Reply:
I think I should make an optimization to avoid filtering the annotations if the `includeAnnotationsIndex` is empty, since in that case none of the csdl annotations will be imported.